File: src-python/doubly_linked_list.py
from typing import TypeVar, Generic


E = TypeVar('E')


class DoublyLinkedList(Generic[E]):

    class _Node:
        def __init__(self, data: E | None):
            self.data: E = data
            self.prev: DoublyLinkedList._Node | None = None
            self.next: DoublyLinkedList._Node | None = None

    def __init__(self):
        self._dummy_head = self._Node(None)
        self._size = 0

    def insert(self, index: int, e: E) -> None:
        if index < 0 or index > self._size:
            raise IndexError("索引越界")

        prev: DoublyLinkedList._Node = self._dummy_head
        for _ in range(0, index):
            prev = prev.next

        new_node: DoublyLinkedList._Node = self._Node(e)
        new_node.prev = prev
        new_node.next = prev.next
        if prev.next is not None:
            prev.next.prev = new_node
        prev.next = new_node
        self._size += 1

    def remove(self, index: int) -> E:
        if index < 0 or index >= self._size:
            raise IndexError("索引越界")

        prev: DoublyLinkedList._Node = self._dummy_head
        for _ in range(0, index):
            prev = prev.next

        del_node: DoublyLinkedList._Node = prev.next
        prev.next = del_node.next
        if del_node.next is not None:
            del_node.next.prev = prev
        del_node.prev = None
        del_node.next = None
        self._size -= 1

        return del_node.data

    def print_doubly_linked_list(self):
        cur: DoublyLinkedList._Node = self._dummy_head.next
        while cur is not None:
            print(cur.data, end="")
            if cur.next is not None:
                print(" <-> ", end="")
            cur = cur.next

if __name__ == "__main__":
    linked_list = DoublyLinkedList[int]()
    linked_list.insert(0, 0)
    linked_list.insert(1, 1)
    linked_list.insert(2, 2)
    linked_list.insert(3, 3)
    linked_list.insert(4, 4)
    linked_list.insert(5, 5)
    linked_list.insert(6, 6)
    linked_list.insert(7, 7)
    linked_list.insert(8, 8)
    linked_list.insert(9, 9)
    linked_list.print_doubly_linked_list()

File: src-java/src/main/java/DoublyLinkedList.java
import org.w3c.dom.Node;

/**
 * ClassName: DoublyLinkedList
 * Package: PACKAGE_NAME
 * Description:
 *
 * @Author AnXin
 * @Create 2026/5/31 16:30
 * @Version 1.0
 */
public class DoublyLinkedList<E> {

    private Node dummyHead;
    private int size;

    DoublyLinkedList() {
        this.dummyHead = new Node(null, null, null);
        this.size = 0;
    }

    private class Node {
        E data;
        Node prev;
        Node next;

        Node() {

        }

        Node(E data) {
            this.data = data;
        }

        Node(E data, Node prev, Node next) {
            this.data = data;
            this.prev = prev;
            this.next = next;
        }
    }

    public void insert(int index, E e) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("索引越界");
        }

        Node prev = dummyHead;
        for (int i = 0; i < index; i++) {
            prev = prev.next;
        }

        Node newNode = new Node(e);
        newNode.prev = prev;
        newNode.next = prev.next;
        if (prev.next != null) {
            prev.next.prev = newNode;
        }
        prev.next = newNode;
        size++;
    }

    public E remove(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("索引越界");
        }

        Node prev = dummyHead;
        for (int i = 0; i < index; i++) {
            prev = prev.next;
        }

        Node delNode = prev.next;
        prev.next = delNode.next;
        if (delNode.next != null) {
            delNode.next.prev = prev;
        }
        delNode.prev = null;
        delNode.next = null;
        size--;

        return delNode.data;
    }

    public void printDoublyLinkedList() {
        Node cur = dummyHead.next;
        while (cur != null) {
            System.out.print(cur.data);
            if (cur.next != null) {
                System.out.print(" <-> ");
            }
            cur = cur.next;
        }
    }
}
